util-images/probes/pkg/dnspropagation/probe.go

```diff
@@ -0,0 +1,221 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dnspropagation
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"math"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+)
+
+var (
+	statefulSet   = flag.String("dns-propagation-probe-stateful-set", "", "Name of the statefulSet workload")
+	service       = flag.String("dns-propagation-probe-service", "", "Name of the headless service that exposes the statefulSet resides")
+	namespace     = flag.String("dns-propagation-probe-namespace", "default", "The namespace where the statefulSet resides")
+	clusterDomain = flag.String("dns-propagation-probe-cluster-domain", "cluster", "Name of cluster domain where the statefulSet resides")
+	suffix        = flag.String("dns-propagation-probe-suffix", "local", "DNS label suffix")
+	interval      = flag.Duration("dns-propagation-probe-interval", 100*time.Millisecond, "Interval between DNS lookups")
+	podCount      = flag.Int("dns-propagation-probe-pod-count", 0, "Number of pods in the statefulSet")
+	sampleCount   = flag.Int("dns-propagation-probe-sample-count", 0, "Number of pods to test dns propagation against in the statefulSet, defaults to min(100, Ceil(SQRT(podCount))")
+)
+
+type DNSPodPropagationResult struct {
+	podName  string
+	duration time.Duration
+}
+
+// Run is the entry function for the probe.
+func Run() {
+	if *statefulSet == "" {
+		klog.Fatal("--dns-propagation-probe-stateful-set has not been set")
+	}
+	if *service == "" {
+		klog.Fatal("--dns-propagation-probe-service-set has not been set")
+	}
+	if *podCount <= 0 {
+		klog.Fatal("--dns-propagation-probe-pod-count has not been set or is not a positive number")
+	}
+	if *sampleCount <= 0 {
+		f := int(math.Ceil(math.Sqrt(float64(*podCount))))
+		f = int(math.Min(float64(f), 100))
+		sampleCount = &f
+		klog.Warningf("dns-propagation-probe-sample-count not set, defaulting to min(100, Ceil(SQRT(%v))= %v", *podCount, *sampleCount)
+	}
+	// creates the in-cluster config
+	kubeConfig, err := rest.InClusterConfig()
+	if err != nil {
+		klog.Fatalf("Can not build inClusterConfig, error:%v", err)
+	}
+	// creates the inCluster kube client
+	clientset, err := kubernetes.NewForConfig(kubeConfig)
+
+	if err != nil {
+		klog.Fatalf("Failed to build kubeClient, error:%v", err)
+	}
+	//TODO deprecated as of Go 1.20. To remove when go version gets bumped
+	rand.Seed(time.Now().UnixNano())
+	runProbe(clientset)
+	for {
+		klog.V(2).Infof("dns propagation probe complete, waiting until the test finishes...")
+		time.Sleep(60 * time.Second)
+	}
+}
+
+// runProbe runs the DNS propagation probe.
+func runProbe(client kubernetes.Interface) {
+	klog.V(1).Infof("DNS propagation probe started")
+	var wg sync.WaitGroup
+	ch := make(chan DNSPodPropagationResult)
```

Comment:
Nitpick: we can increase the channel's capacity to match the number of goroutines so we do not block on sending data through it.

Reply:
DONE